index.js

```diff
@@ -20,39 +20,38 @@
 //. purpose: the [`typeof`][1] operator and [`Object.prototype.toString`][2].
 //. Each has pros and cons, but neither supports user-defined types.
 //.
-//. This package specifies an [algorithm][3] for deriving a _type identifier_
-//. from any JavaScript value, and exports an implementation of the algorithm.
-//. Authors of algebraic data types may follow this specification in order to
-//. make their data types compatible with the algorithm.
+//. This package exports a function which derives a _type identifier_ from any
+//. JavaScript value, and a [specification][3] for customizing the type
+//. identifier of a value.
 //.
-//. ### Algorithm
+//. ### Specification
 //.
-//. 1.  Take any JavaScript value `x`.
+//. For a type to be compatible with the algorithm:
 //.
-//. 2.  If `x` is `null` or `undefined`, go to step 6.
-//.
-//. 3.  If `x.constructor` evaluates to `null` or `undefined`, go to step 6.
-//.
-//. 4.  If `x.constructor.prototype === x`, go to step 6. This check prevents a
-//.     prototype object from being considered a member of its associated type.
+//.   - every member of the type MUST have a `constructor` property pointing
+//.     to an object known as the _type representative_;
 //.
-//. 5.  If `typeof x.constructor['@@type']` evaluates to `'string'`, return
-//.     the value of `x.constructor['@@type']`.
+//.   - the type representative MUST have a `@@type` property;
 //.
-//. 6.  Return the [`Object.prototype.toString`][2] representation of `x`
-//.     without the leading `'[object '` and trailing `']'`.
+//.   - the type representative's `@@type` property (the _type identifier_)
+//.     MUST be a string primitive and SHOULD have format:
+//.     `'<namespace>/<name>[@<version>]'`, where:
 //.
-//. ### Compatibility
+//.       - The `namespace` MUST consist of one or more non-`/` characters, MAY
+//.         contain a single `/` if it starts with `@` (to support
+//.         [npm scopes][4]), and SHOULD equal the npm package name which
+//.         defines the type; and
```

Comment:
> My goal is not for every valid `namespace` [to] be a valid npm package name, but for every valid npm package name to be a valid `namespace`.

This is *exactly* how it's currently implemented, and how I see it too. I think we have the same goal, but I think you are contradicting yourself in stating that *If namespace begins with `@` it MUST be followed, in order, by one or more non-`/` characters, a `/`, and one or more non-`/` characters.*

> `@abc` should be invalid.

If we look at the definition of `<package>` (the second possible valid `<namespace>`):

```txt
<package>           ::= <non-empty>
<non-empty>         ::= <any> [ <non-empty> ]
<any>               ::= / | <non-slash>
<non-slash>         ::= A | B | C | ...
```

Then we may conclude that `@abc` is actually a *valid* <package> (assuming `<non-slash>` includes all possible characters, giving `<any>` its name).

If you wish for `@abc` to be *invalid*, some arbitrary rule has to be introduced, like "the namespace may not begin with `@`, or contain `@` characters".